# Multilayer Perceptron and Neural Network

---

## SmartSeq

In this first section, we will utilise the MLPClassifier class in sklearn. We will start with the SmartSeq datasets and then move on to the larger DropSeq datasets.

### General Imports

In [2]:
import numpy as np
import torch.nn as nn
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.neural_network import MLPClassifier

In [5]:
#Training sets
df_HCC = pd.read_csv("raw_data/HCC1806_SmartS_Filtered_Normalised_3000_Data_train.txt", delimiter="\ ",engine='python',index_col=0)
df_MCF = pd.read_csv("raw_data/MCF7_SmartS_Filtered_Normalised_3000_Data_train.txt", delimiter="\ ",engine='python',index_col=0)

#Metadata files
df_meta_HCC = pd.read_csv("raw_data/HCC1806_SmartS_MetaData.tsv",delimiter="\t",engine='python',index_col=0)
df_meta_MCF = pd.read_csv("raw_data/MCF7_SmartS_MetaData.tsv",delimiter="\t",engine='python',index_col=0)

#Transposing
df_HCC = df_HCC.T
df_MCF = df_MCF.T

print(df_HCC.shape)
df_HCC.head()


(182, 3000)


"DDIT4"  "ANGPTL4"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....        0         48   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...     8739       2101   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...    13098      14032   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...     2880        356   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...     7777       5661   

                                                    "CALML5"  "KRT14"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....         0      321   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...        55       96   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...         0        0   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...         0     6211   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...      4383        0   

                                                    "CCNB1"  "IGFBP3"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....      298        82   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...     1824      1938   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...     1616       247   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...        3      3430   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...      145      4618   

                                                    "AKR1C2"  "KRT6A"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....      6250      634   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...        62        0   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...       430      907   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...        79     1953   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...       246       85   

                                                    "NDRG1"  "KRT4"  ...  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....        0       0  ...   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...      522     413  ...   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...      348       0  ...   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...      592     176  ...   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...      206       0  ...   

                                                    "MST1R"  "ZYG11A"  "NRG1"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....       78        10     136   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...      279         0     264   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...      311         0      38   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...      125         0      16   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...      268         1      25   

                                                    "RBMS3"  "VCPIP1"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....        0         0   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...        0       134   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...        0         0   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...        0         4   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...        0         0   

                                                    "LINC02693"  "OR8B9P"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....           29         0   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...           68         0   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...            0         0   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...            1         0   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...            0         0   

                                                    "NEAT1"  "ZDHHC23"  \
"output.STAR.PCRPlate1G12_Normoxia_S32_Aligned....       29          0   
"output.STAR.PCRPlate1G1_Hypoxia_S102_Aligned.s...      213          0   
"output.STAR.PCRPlate1G2_Hypoxia_S2_Aligned.sor...       92          0   
"output.STAR.PCRPlate1G3_Hypoxia_S7_Aligned.sor...        1          0   
"output.STAR.PCRPlate1G4_Hypoxia_S107_Aligned.s...      128          0   


In [45]:
print(df_MCF.shape)
df_MCF.head()


(250, 3000)


"CYP1B1"  "CYP1B1-AS1"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...       343           140   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...       131            59   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...       452           203   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...        27             7   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...      5817          2669   

                                                    "CYP1A1"  "NDRG1"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...         0        0   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...         0        1   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...         0        0   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...         0        0   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...         0      654   

                                                    "DDIT4"  "PFKFB3"  "HK2"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...      386        75      0   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...      289        42      0   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...        0         0      0   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...      288       214     15   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...     2484      1603    748   

                                                    "AREG"  "MYBL2"  "ADM"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...       0      476      0   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...     856     1586      0   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...       0      775      0   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...     242     1191      0   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...     436        0      0   

                                                    "CDC20"  "KRT80"  "PLK1"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...      250        0      57   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...      125      407       0   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...      219       53     157   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...      261      307     184   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...        0     1701       0   

                                                    "TK1"  "GDF15"  "TFF1"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...    270      541      80   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...    938        0    1894   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...   2068       28     662   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...    519       44    1297   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...      0      996    1007   

                                                    "KIF20A"  "VEGFA"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...        40        4   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...         1       56   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...       175        0   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...        82       80   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...         0      496   

                                                    "BCAS1"  "RASD1"  "NR4A1"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...        0      127      370   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...       21       10      235   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...        1       33       32   
"output.STAR.2_B6_Norm_S60_Aligned.sortedByCoor...       21        0      334   
"output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoor...      255       22        5   

                                                    "IGFBP3"  "BNIP3"  \
"output.STAR.2_B3_Norm_S57_Aligned.sortedByCoor...         0       34   
"output.STAR.2_B4_Norm_S58_Aligned.sortedByCoor...         0       33   
"output.STAR.2_B5_Norm_S59_Aligned.sortedByCoor...         0       38   
"output.STAR.2_B6_Norm_S60

### Labelling



In [4]:
def Label(df, meta):
    Names = [cell[1:-1] for cell in df.index]
    Hypo = [meta.loc[i,"Condition"] for i in Names]
    OH_Hypo = [1 if x == "Normo" else 0 for x in Hypo]
    result = pd.DataFrame(OH_Hypo, columns=["Condition"], index=df.index)
    return result

We use the label function to access the meta file and obtain the classificiation of the cells.

### K-fold Cross Validation


In [34]:
def Kfold_crossval(X, Y, clf, n_folds = 5):
    kf = KFold(n_splits= n_folds)
    for i, (train_index, test_index) in enumerate(kf.split(X)):
        X_train = X.iloc[train_index]
        Y_train = Y.iloc[train_index]
        X_test = X.iloc[test_index]
        Y_test = Y.iloc[test_index]
        if i == 0:
            clf.fit(X_train,Y_train.Condition.values)
        else:
            clf.partial_fit(X_train,Y_train.Condition.values)
    return clf

This function trains the MLP using the fit and partial fit methods. The training set is split into 5 parts and we implement the fitting method 5 times (as number of folds is equal to five). Then, after fitting, it return the MLP model.

### Train Test Random Split


In [3]:
def split_train_test(data, test_ratio):
    shuffled_indices = np.random.permutation(len(data))
    test_set_size = int(len(data) * test_ratio)
    test_indices = shuffled_indices[:test_set_size]
    train_indices = shuffled_indices[test_set_size:]
    return data.iloc[train_indices], data.iloc[test_indices]

Here, we split the dataset into our training set and test set. By default, our test_ratio is 0.2.

### Testing

In [48]:
def Test_Accuracy(df,meta,n_simul):
    score = []
    #clf = MLPClassifier(random_state=1, max_iter=300)
    clf = MLPClassifier(hidden_layer_sizes=250,max_iter = 1000, random_state=1)
    for i in range(n_simul):
        df_train , df_test = split_train_test(df, 0.2)
        Y_train = Label(df_train, meta)
        Y_test = Label(df_test, meta)
        clf.fit(df_train,Y_train.Condition.values)
        score.append(clf.score(df_test,Y_test.Condition.values))
    mean_score = sum(score)/n_simul
    return mean_score

n_simul = 200

score_HCC = Test_Accuracy(df_HCC,df_meta_HCC,n_simul)
score_MCF = Test_Accuracy(df_MCF,df_meta_MCF,n_simul)

print("Mean accuracy HCC:",score_HCC * 100, "%")
print("Mean accuracy MCF:",score_MCF * 100,"%")

Mean accuracy HCC: 96.37500000000016 %
Mean accuracy MCF: 100.0 %


To get an idea of how well our MLP perfoms, we run a simulation. Each time, the train and test set is split randomly. Therefore, each time, the MLP is trained and tested on the randomly divided sets. By taking the average of the scores across the trials, we get an idea of our MLP's performance. For now, we are using just the normal .fit method and not our Kfold cross validation training method. We observe that the MLP performs perfectly on the MCF dataset while it has a very high accuracy on the HCC dataset of 96.38%.

### Simulation with K fold cross validation

In [49]:
def Kfold_Test_Accuracy(df,meta,n_simul):
    score = []
    for i in range(n_simul):
        df_train , df_test = split_train_test(df, 0.2)
        Y_train = Label(df_train, meta)
        Y_test = Label(df_test, meta)
        clf_return = Kfold_crossval(df_train,Y_train, clf = MLPClassifier(hidden_layer_sizes=250,max_iter = 1000, random_state=1))
        score.append(clf_return.score(df_test,Y_test.Condition.values))
    mean_score = sum(score)/n_simul
    return mean_score

n_simul = 100

score_HCC = Kfold_Test_Accuracy(df_HCC,df_meta_HCC,n_simul)
score_MCF = Kfold_Test_Accuracy(df_MCF,df_meta_MCF,n_simul)

print("Mean accuracy HCC:",score_HCC * 100, "%")
print("Mean accuracy MCF:",score_MCF * 100,"%")

Mean accuracy HCC: 97.1944444444445 %
Mean accuracy MCF: 100.0 %


Using the Kfold training method written before, we our able to train our MLP better. Noticeably, the accuracy on the HCC dataset has increased while the accuracy on the MCF dataset remains a perfect 100%.

---
## DropSeq

Now, we are dealing with a much larger dataset. Therefore, one can expect a drop in accuracy (as well as a significant increase in running time)

In [6]:
df_HCC_drop = pd.read_csv("HCC1806_Filtered_Normalised_3000_Data_train.txt", delimiter="\ ",engine='python',index_col=0)
df_MCF_drop = pd.read_csv("MCF7_Filtered_Normalised_3000_Data_train.txt", delimiter="\ ",engine='python',index_col=0)

df_HCC_drop = df_HCC_drop.T
df_MCF_drop = df_MCF_drop.T



In [9]:
print(df_HCC_drop.shape)
df_HCC_drop.head()


(14682, 3000)


"H1-5"  "MALAT1"  "MT-RNR2"  "ARVCF"  "BCYRN1"  \
"AAAAAACCCGGC_Normoxia"       2         3          0        0         0   
"AAAACCGGATGC_Normoxia"       2         3          0        0         1   
"AAAACGAGCTAG_Normoxia"       5         2          0        0         1   
"AAAACTTCCCCG_Normoxia"       1         3          0        0         0   
"AAAAGCCTACCC_Normoxia"       0        12          0        0         0   

                         "ATXN7L2"  "IGFBP3"  "H1-3"  "CTIF"  "RNF123"  \
"AAAAAACCCGGC_Normoxia"          0         0       2       0         0   
"AAAACCGGATGC_Normoxia"          0         0       0       0         0   
"AAAACGAGCTAG_Normoxia"          0         1       3       0         0   
"AAAACTTCCCCG_Normoxia"          0         0       0       0         0   
"AAAAGCCTACCC_Normoxia"          0         0       0       0         0   

                         "CCDC97"  "RPL12P42"  "TRPT1"  "PARP12"  "ABTB2"  \
"AAAAAACCCGGC_Normoxia"         0           0        0         0        0   
"AAAACCGGATGC_Normoxia"         0           0        0         0        0   
"AAAACGAGCTAG_Normoxia"         0           0        0         0        0   
"AAAACTTCCCCG_Normoxia"         0           0        0         0        0   
"AAAAGCCTACCC_Normoxia"         0           0        0         0        0   

                         "CRLS1"  "SYCE1L"  "REP15"  "NDST2"  "TBC1D22A-AS1"  \
"AAAAAACCCGGC_Normoxia"        0         0        0        0               0   
"AAAACCGGATGC_Normoxia"        0         0        0        0               0   
"AAAACGAGCTAG_Normoxia"        0         0        0        0               0   
"AAAACTTCCCCG_Normoxia"        0         0        0        0               0   
"AAAAGCCTACCC_Normoxia"        0         0        0        0               0   

                         "GBP2"  "PRKAA2"  "H2AC11"  "RAPGEF3"  "CARD9"  \
"AAAAAACCCGGC_Normoxia"       0         0         1          0        0   
"AAAACCGGATGC_Normoxia"       0         0         1          0        0   
"AAAACGAGCTAG_Normoxia"       0         0         0          0        0   
"AAAACTTCCCCG_Normoxia"       0         0         0          0        0   
"AAAAGCCTACCC_Normoxia"       0         0         0          0        0   

                         "MARCHF3"  "EGFLAM"  "ZNF710"  "H4C3"  "ANXA6"  \
"AAAAAACCCGGC_Normoxia"          0         0         0       2        0   
"AAAACCGGATGC_Normoxia"          0         0         0       9        0   
"AAAACGAGCTAG_Normoxia"          0         0         0       7        0   
"AAAACTTCCCCG_Normoxia"          0         0         0       2        0   
"AAAAGCCTACCC_Normoxia"          0         0         0       1        0   

                         "CTF1"  "C12orf50"  "RRAGD"  "ZSCAN18"  "DDIT4"  \
"AAAAAACCCGGC_Normoxia"       0           0        0          0        0   
"AAAACCGGATGC_Normoxia"       0           0        0          0        0   
"AAAACGAGCTAG_Normoxia"       0           0        0          0        0   
"AAAACTTCCCCG_Normoxia"       0           0        0          0        0   
"AAAAGCCTACCC_Normoxia"       0           0        0          0        0   

                         "HSPA5"  "MB"  "TMPRSS13"  "ZNRF2"  "H1-2"  "MT-TV"  \
"AAAAAACCCGGC_Normoxia"        0     0           0        0       0        0   
"AAAACCGGATGC_Normoxia"        1     0           0        0       1        0   
"AAAACGAGCTAG_Normoxia"        0     0           0        0       1        0   
"AAAACTTCCCCG_Normoxia"        1     0           0        0       1        0   
"AAAAGCCTACCC_Normoxia"        2     0           0        0       0        0   

                         "UFSP1"  "PLEC"  "EXOSC10-AS1"  "NAT14"  "RASSF10"  \
"AAAAAACCCGGC_Normoxia"        0       0              0        0          0   
"AAAACCGGATGC_Normoxia"        0       0              0        0          0   
"AAAACGAGCTAG_Normoxia"        0       0              0        0          0   
"AAAACTTCCCCG_Normoxia"      

In [10]:
print(df_MCF_drop.shape)
df_MCF_drop.head()

(21626, 3000)


"MALAT1"  "MT-RNR2"  "NEAT1"  "H1-5"  "TFF1"  \
"AAAAACCTATCG_Normoxia"         1          0        0       0       4   
"AAAACAACCCTA_Normoxia"         3          0        0       0       1   
"AAAACACTCTCA_Normoxia"         3          0        0       0       1   
"AAAACCAGGCAC_Normoxia"         6          2        0       0       1   
"AAAACCTAGCTC_Normoxia"         4          0        0       0       0   

                         "MT-RNR1"  "H4C3"  "GDF15"  "KRT81"  "MT-CO3"  \
"AAAAACCTATCG_Normoxia"          0       0        0        0         0   
"AAAACAACCCTA_Normoxia"          0       1        0        0         0   
"AAAACACTCTCA_Normoxia"          0       0        0        0         0   
"AAAACCAGGCAC_Normoxia"          0       1        0        0         0   
"AAAACCTAGCTC_Normoxia"          0       0        0        0         1   

                         "C2orf42"  "SLC4A11"  "MT-ND4"  "DNAI4"  "MT-ND3"  \
"AAAAACCTATCG_Normoxia"          0          0         0        0         0   
"AAAACAACCCTA_Normoxia"          0          0         0        0         0   
"AAAACACTCTCA_Normoxia"          0          0         0        0         0   
"AAAACCAGGCAC_Normoxia"          0          0         3        0         0   
"AAAACCTAGCTC_Normoxia"          0          0         1        0         1   

                         "MT-CYB"  "H1-3"  "IGFBP5"  "PCSK5"  "TMSB10"  \
"AAAAACCTATCG_Normoxia"         0       2         0        0         0   
"AAAACAACCCTA_Normoxia"         0       0         0        0         0   
"AAAACACTCTCA_Normoxia"         0       0         0        0         0   
"AAAACCAGGCAC_Normoxia"         0       1         0        0         0   
"AAAACCTAGCTC_Normoxia"         0       0         0        0         0   

                         "CMTM3"  "MT-ATP6"  "MT-CO2"  "GCNA"  "SCMH1"  \
"AAAAACCTATCG_Normoxia"        0          0         0       0        0   
"AAAACAACCCTA_Normoxia"        0          1         0       0        0   
"AAAACACTCTCA_Normoxia"        0          2         0       0        0   
"AAAACCAGGCAC_Normoxia"        0          0         0       0        0   
"AAAACCTAGCTC_Normoxia"        0          0         1       0        0   

                         "IFI27"  "ZNF304"  "PLEKHS1"  "ZDHHC8"  "MT-TS1"  \
"AAAAACCTATCG_Normoxia"        0         0          0         0         2   
"AAAACAACCCTA_Normoxia"        0         0          0         0         0   
"AAAACACTCTCA_Normoxia"        0         0          0         0         0   
"AAAACCAGGCAC_Normoxia"        0         0          0         0         0   
"AAAACCTAGCTC_Normoxia"        0         0          0         0         0   

                         "SRP14-DT"  "ZBTB49"  "NKAIN1"  "MTATP6P1"  "MT-ND6"  \
"AAAAACCTATCG_Normoxia"           0         0         0           0         0   
"AAAACAACCCTA_Normoxia"           0         0         0           0         0   
"AAAACACTCTCA_Normoxia"           0         0         0           0         1   
"AAAACCAGGCAC_Normoxia"           0         0         0           0         0   
"AAAACCTAGCTC_Normoxia"           0         0         0           0         0   

                         "MT-CO1"  "MT-ND2"  "MTND1P23"  "RCOR2"  "MT-TV"  \
"AAAAACCTATCG_Normoxia"         2         0           0        0        0   
"AAAACAACCCTA_Normoxia"         0         0           0        0        0   
"AAAACACTCTCA_Normoxia"         0         0           0        0        0   
"AAAACCAGGCAC_Normoxia"         0         0           0        0        0   
"AAAACCTAGCTC_Normoxia"         0         0           0        0        0   

                         "MT-ND4L"  "PAQR4"  "LPAR6"  "MT-ND5"  "MT-TN"  \
"AAAAACCTATCG_Normoxia"          0        0        0         0        0   
"AAAACAACCCTA_Normoxia"          0        0        0         0        0   
"AAAACACTCTCA_Normoxia"          0        0        0         0        0   
"AAAACCAGGCAC_Normoxia"          0        0        0         1       

### Create Label dataframes

In [7]:
def Label_drop(df):
    Names = [cell[1:-1] for cell in df.index]
    OH_Hypo = [1 if "Normoxia" in x else 0 for x in Names]
    result = pd.DataFrame(OH_Hypo, columns=["Condition"], index = df.index)
    return result

Condition_HCC = Label_drop(df_HCC_drop)
Condition_MCF = Label_drop(df_MCF_drop)

Since we don't have the meta files for the DropSeq datasets, we use the names of the indices (cells) to classify them.

### Test Accuracy DropSeq

In [50]:
def DropSeq_Test_Accuracy(df,n_simul):
    score = []
    #clf = MLPClassifier(random_state=1, max_iter=300)
    clf = MLPClassifier(hidden_layer_sizes=225,max_iter = 500, random_state=1)
    for i in range(n_simul):
        df_train , df_test = split_train_test(df, 0.2)
        Y_train = Label_drop(df_train)
        Y_test = Label_drop(df_test)
        clf.fit(df_train,Y_train.Condition.values)
        score.append(clf.score(df_test,Y_test.Condition.values))
    mean_score = sum(score)/n_simul
    return mean_score

n_simul = 5
score_HCC = DropSeq_Test_Accuracy(df_HCC_drop,n_simul)
print("Mean accuracy HCC:",score_HCC * 100, "%")

score_MCF = DropSeq_Test_Accuracy(df_MCF_drop,n_simul)
print("Mean accuracy MCF:",score_MCF * 100, "%")


Mean accuracy HCC: 95.17711171662124 %
Mean accuracy MCF: 97.98381502890173 %


We use the same function as before except this time on DropSeq. There is a drop in accuracy but not too much. 

### K fold Cross Validation on DropSeq Model

In [38]:
def Drop_Kfold_Test_Accuracy(df,n_simul):
    score = []
    for i in range(n_simul):
        df_train , df_test = split_train_test(df, 0.2)
        Y_train = Label_drop(df_train)
        Y_test = Label_drop(df_test)
        Y = Label_drop(df)
        clf_return = Kfold_crossval(df_train,Y_train, clf = MLPClassifier(hidden_layer_sizes=225,max_iter = 500, random_state=1))
        score.append(clf_return.score(df_test,Y_test))
    mean_score = sum(score)/n_simul
    return mean_score

n_simul = 25

score_HCC = Drop_Kfold_Test_Accuracy(df_HCC_drop,n_simul)
print("Mean accuracy HCC:",score_HCC * 100, "%")

score_MCF = Drop_Kfold_Test_Accuracy(df_MCF_drop,n_simul)
print("Mean accuracy MCF:",score_MCF * 100, "%")

Mean accuracy HCC: 95.19209809264305 %
Mean accuracy MCF: 97.85618497109826 %


The Same Kfold Cross Validation function is used for the DropSeq datasets. We observe that the accuracy does not improve much. This concludes the first section of Multilayer Perceptron, we shall now move on to Neural Networks and hyperparameter optimization. 

---
---
## Neural Network Hyperparameter tuning

So far, we have obtained "high" accuracy with our basic MLP classifier. But, can we do better? In this section, we will implement neural networks and we try to find the optimal hyperparameters for our datasets. We will use the keras library to implement our neural networks and the bayesian optimisation library to optimise the hyperparameters and layers.

### General Imports

In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, Dropout
from keras.optimizers import Adam, SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from keras.callbacks import EarlyStopping, ModelCheckpoint
from keras.wrappers.scikit_learn import KerasClassifier

from math import floor
from sklearn.metrics import make_scorer, accuracy_score
from bayes_opt import BayesianOptimization
from sklearn.model_selection import StratifiedKFold
import pickle
from keras.layers import LeakyReLU
LeakyReLU = LeakyReLU(alpha=0.1)

import warnings
warnings.filterwarnings('ignore')
pd.set_option("display.max_columns", None)

### HCC

In [9]:
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

# Load dataset
trainSet_HCC_drop = df_HCC_drop

In [10]:
# HCC Validation split
HCC_train_drop, HCC_test_drop = split_train_test(trainSet_HCC_drop, 0.2)
HCC_y_train_drop, HCC_y_test_drop = Label_drop(HCC_train_drop), Label_drop(HCC_test_drop)

#### Tuning the Hyperparameters

The first hyperparameter to optimize is the number of neurons in each hidden layer. In this case, the number of neurons in each layer is set equal. The number of neurons should be adjusted according to the complexity of the solution. Tasks involving more complex prediction layers will require more neurons. 

The activation function is a parameter for each layer. Input data is fed to the input layer, then to the hidden layers, and finally to the output layer. The output layer contains the output values. An input value that changes from one level to another always changes according to the activation function. The activation function determines how the level's input values ​​are transformed into output values. Output values ​​of one level are passed as input values ​​to the next level. Then the values ​​are recomputed at the next level to produce another level of output values. Here, we have nine activation functions. Each activation function has its formula (and graph) for calculating input values. 

Neural network layers are compiled and assigned to the optimizer. The optimizer is responsible for changing the learning rate and weights of neurons in the neural network to achieve minimal loss performance. Optimizers are very important to achieve the highest possible accuracy or lowest possible loss. You can choose from seven optimizers. Everyone has different concepts. 

One of the optimizer hyperparameters is the learning rate. The learning rate controls the step size required before the model works with minimal loss. The higher the learning rate, the faster the model learns, but it may miss the minimum loss function and reach only its neighbors. The lower the learning rate, the more likely it is to find a function with minimal loss. On the other hand, a lower learning rate requires higher epochs or more resources of time and memory capacity. 

Batch size is the number of training data sub-samples for the input. The smaller batch size makes the learning process faster, but the variance of the validation dataset accuracy is higher. A bigger batch size has a slower learning process, but the validation dataset accuracy has a lower variance.

The number of times a whole dataset is passed through the neural network model is called an epoch. One epoch means that the training dataset is passed forward and backward through the neural network once. A too-small number of epochs results in underfitting because the neural network has not learned much enough. The training dataset needs to pass multiple times or multiple epochs are required. On the other hand, too many epochs will lead to overfitting where the model can predict the data very well, but cannot predict new unseen data well enough. The number of epoch must be tuned to gain the optimal result.

#### Tuning the Layers

Layers in Neural Networks also determine the result of the prediction model. A smaller number of layers is enough for a simpler problem, but a larger number of layers is needed to build a model for a more complicated problem.

Inserting regularization layers in a neural network can help prevent overfitting. This demonstration tries to tune whether to add regularization layers or not. There are two regularization layers to use here.

Batch normalization is placed after the first hidden layers. The batch normalization layer normalizes the values passed to it for every batch. This is similar to the standard scaler in conventional Machine Learning.

Another regularization layer is the Dropout layer. The dropout layer, as its name suggests, randomly drops a certain number of neurons in a layer. The dropped neurons are not used anymore. The rate of how much percentage of neurons drop is set in the dropout rate.

The following is the code to tune the hyperparameters and layers.

In [ ]:
# Create function
def nn_cl_bo2_HCC_drop(neurons, activation, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate):
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
                 'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
                 'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
                 'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
        
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']
        
    neurons = round(neurons)
    activation = activationL[round(activation)]
    optimizer = optimizerD[optimizerL[round(optimizer)]]
    batch_size = round(batch_size)
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)
        
    def nn_cl_fun():
        nn = Sequential()
        nn.add(Dense(neurons, input_dim=3000, activation=activation))
        if normalization > 0.5:
            nn.add(BatchNormalization())
        for i in range(layers1):
            nn.add(Dense(neurons, activation=activation))
        if dropout > 0.5:
            nn.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            nn.add(Dense(neurons, activation=activation))
        nn.add(Dense(1, activation='sigmoid'))
        nn.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
        return nn
        
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
    nn = KerasClassifier(build_fn=nn_cl_fun, epochs=epochs, batch_size=batch_size, verbose=0)
    
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, HCC_train_drop, HCC_y_train_drop, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    
    return score


In [ ]:
params_nn2 ={
    'neurons': (3000, 100),
    'activation':(0, 9),
    'optimizer':(0,7),
    'learning_rate':(0.01, 1),
    'batch_size':(200, 1000),
    'epochs':(20, 100),
    'layers1':(1,3),
    'layers2':(1,3),
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}

# Run Bayesian Optimization
nn_bo_HCC_drop_layers = BayesianOptimization(nn_cl_bo2_HCC_drop, params_nn2, random_state=111)
nn_bo_HCC_drop_layers.maximize(init_points=25, n_iter=4)

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------
74/74 [==============================] - 2s 24ms/step
| 1         | 0.6051    | 5.51      | 335.3     | 0.4361    | 0.2308    | 43.63     | 1.298     | 1.045     | 0.426     | 2.308e+03 | 0.3377    | 6.935     |
74/74 [==============================] - 0s 4ms/step
| 2         | 0.5826    | 2.14      | 265.0     | 0.6696    | 0.1864    | 41.94     | 1.932     | 1.237     | 0.08322   | 387.8     | 0.794     | 5.884     |
74/74 [==============================] - 0s 5ms/step
| 3         | 0.6051    | 7.337     | 992.8     | 0.5773    | 0.2441    | 53.71     | 1.055     | 1.908     | 0.1143    | 630.1     | 0.6977    | 3.957     |
74/74 [==============================] - 2s 26ms/s

ValueError: LBFGSB - one of the lower bounds is greater than an upper bound.

In [11]:
#| 6         | 0.9594    | 0.3436    | 242.5     | 0.128     | 0.01001   | 38.11     | 2.088     | 1.357     | 0.1876    | 2.566e+03 | 0.683     | 3.283     |
#|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |

#params_nn_HCC_drop_layers = nn_bo_HCC_drop_layers.max['params']

params_nn_HCC_drop_layers = {
    'activation': 0.3436,
    'batch_size': 242,
    'dropout': 0.1279608146375314,
    'dropout_rate': 0.010014947837176436,
    'epochs': 38,
    'layers1': 2,
    'layers2': 1,
    'learning_rate': 0.18755133602737314,
    'neurons': 2566,
    'normalization': 0.6830131255680931,
    'optimizer': 3.283
}

learning_rate = params_nn_HCC_drop_layers['learning_rate']
activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
               'elu', 'exponential', LeakyReLU,'relu']
params_nn_HCC_drop_layers['activation'] = activationL[round(params_nn_HCC_drop_layers['activation'])]

params_nn_HCC_drop_layers['batch_size'] = round(params_nn_HCC_drop_layers['batch_size'])
params_nn_HCC_drop_layers['epochs'] = round(params_nn_HCC_drop_layers['epochs'])
params_nn_HCC_drop_layers['layers1'] = round(params_nn_HCC_drop_layers['layers1'])
params_nn_HCC_drop_layers['layers2'] = round(params_nn_HCC_drop_layers['layers2'])
params_nn_HCC_drop_layers['neurons'] = round(params_nn_HCC_drop_layers['neurons'])

optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','Adam']
optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
             'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
             'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
             'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
params_nn_HCC_drop_layers['optimizer'] = optimizerD[optimizerL[round(params_nn_HCC_drop_layers['optimizer'])]]

params_nn_HCC_drop_layers

{'activation': 'relu',
 'batch_size': 242,
 'dropout': 0.1279608146375314,
 'dropout_rate': 0.010014947837176436,
 'epochs': 38,
 'layers1': 2,
 'layers2': 1,
 'learning_rate': 0.18755133602737314,
 'neurons': 2566,
 'normalization': 0.6830131255680931,
 'optimizer': <keras.optimizers.legacy.adadelta.Adadelta at 0x1dca63feda0>}

We test our neural network on the validation set, now with our optimized hyperparameters and layers. We fit the model to the training set and then make our predictions on the test set.

In [12]:
def nn_cl_fun_HCC_drop_layers():
    nn = Sequential()
    nn.add(Dense(params_nn_HCC_drop_layers['neurons'], input_dim=3000, activation=params_nn_HCC_drop_layers['activation']))
    if params_nn_HCC_drop_layers['normalization'] > 0.5:
        nn.add(BatchNormalization())
    for i in range(params_nn_HCC_drop_layers['layers1']):
        nn.add(Dense(params_nn_HCC_drop_layers['neurons'], activation=params_nn_HCC_drop_layers['activation']))
    if params_nn_HCC_drop_layers['dropout'] > 0.5:
        nn.add(Dropout(params_nn_HCC_drop_layers['dropout_rate'], seed=123))
    for i in range(params_nn_HCC_drop_layers['layers2']):
        nn.add(Dense(params_nn_HCC_drop_layers['neurons'], activation=params_nn_HCC_drop_layers['activation']))
    nn.add(Dense(1, activation='sigmoid'))
    nn.compile(loss='binary_crossentropy', optimizer=params_nn_HCC_drop_layers['optimizer'], metrics=['accuracy'])
    return nn

es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
nn = KerasClassifier(build_fn=nn_cl_fun_HCC_drop_layers, epochs=params_nn_HCC_drop_layers['epochs'], batch_size=params_nn_HCC_drop_layers['batch_size'],
                         verbose=0)
 
nn.fit(HCC_train_drop, HCC_y_train_drop, verbose=1)

Epoch 1/38
49/49 [==============================] - 18s 343ms/step - loss: 0.3143 - accuracy: 0.8698
Epoch 2/38
49/49 [==============================] - 17s 343ms/step - loss: 0.0707 - accuracy: 0.9765
Epoch 3/38
49/49 [==============================] - 16s 330ms/step - loss: 0.0223 - accuracy: 0.9955
Epoch 4/38
49/49 [==============================] - 16s 326ms/step - loss: 0.0068 - accuracy: 0.9992
Epoch 5/38
49/49 [==============================] - 16s 329ms/step - loss: 0.0032 - accuracy: 0.9999
Epoch 6/38
49/49 [==============================] - 16s 327ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 7/38
49/49 [==============================] - 16s 332ms/step - loss: 8.1958e-04 - accuracy: 1.0000
Epoch 8/38
49/49 [==============================] - 16s 319ms/step - loss: 5.2013e-04 - accuracy: 1.0000
Epoch 9/38
49/49 [==============================] - 16s 327ms/step - loss: 4.7963e-04 - accuracy: 1.0000
Epoch 10/38
49/49 [==============================] - 16s 328ms/step - loss: 3.4

In [14]:
accuracy = accuracy_score(HCC_y_test_drop, nn.predict(HCC_test_drop))

print(accuracy)

92/92 [==============================] - 4s 39ms/step
0.9560626702997275


The neural network gives us an average of 95.61% accuracy on the test set. Albeit small, it is still an improvement from our standard MLP classifier. This concludes our neural network optimization part on DropSeq HCC. We shall now move on to DropSeq MCF.

---

### MCF

In [15]:
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

# Load dataset
trainSet_MCF_drop = df_MCF_drop

In [16]:
# MCF Validation split
MCF_train_drop, MCF_test_drop = split_train_test(trainSet_MCF_drop, 0.2)
MCF_y_train_drop, MCF_y_test_drop = Label_drop(MCF_train_drop), Label_drop(MCF_test_drop)

In [ ]:
# Create function
def nn_cl_bo2_MCF_drop(neurons, activation, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate):
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
                 'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
                 'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
                 'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
        
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']
        
    neurons = round(neurons)
    activation = activationL[round(activation)]
    optimizer = optimizerD[optimizerL[round(optimizer)]]
    batch_size = round(batch_size)
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)
        
    def nn_cl_fun():
        nn = Sequential()
        nn.add(Dense(neurons, input_dim=3000, activation=activation))
        if normalization > 0.5:
            nn.add(BatchNormalization())
        for i in range(layers1):
            nn.add(Dense(neurons, activation=activation))
        if dropout > 0.5:
            nn.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            nn.add(Dense(neurons, activation=activation))
        nn.add(Dense(1, activation='sigmoid'))
        nn.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
        return nn
        
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
    nn = KerasClassifier(build_fn=nn_cl_fun, epochs=epochs, batch_size=batch_size, verbose=0)
    
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, MCF_train_drop, MCF_y_train_drop, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    
    return score


In [ ]:
params_nn2 ={
    'neurons': (3000, 100),
    'activation':(0, 9),
    'optimizer':(0,7),
    'learning_rate':(0.01, 1),
    'batch_size':(200, 1000),
    'epochs':(20, 100),
    'layers1':(1,3),
    'layers2':(1,3),
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}

# Run Bayesian Optimization
nn_bo_MCF_drop_layers = BayesianOptimization(nn_cl_bo2_MCF_drop, params_nn2, random_state=111)
nn_bo_MCF_drop_layers.maximize(init_points=25, n_iter=4)

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------
109/109 [==============================] - 6s 54ms/step
| 1         | 0.5841    | 5.51      | 335.3     | 0.4361    | 0.2308    | 43.63     | 1.298     | 1.045     | 0.426     | 2.308e+03 | 0.3377    | 6.935     |
109/109 [==============================] - 2s 17ms/step
| 2         | 0.5921    | 2.14      | 265.0     | 0.6696    | 0.1864    | 41.94     | 1.932     | 1.237     | 0.08322   | 387.8     | 0.794     | 5.884     |
109/109 [==============================] - 3s 22ms/step
| 3         | 0.4159    | 7.337     | 992.8     | 0.5773    | 0.2441    | 53.71     | 1.055     | 1.908     | 0.1143    | 630.1     | 0.6977    | 3.957     |
109/109 [==============================] -

ValueError: Input y contains NaN.

In [17]:
#params_nn_MCF_drop_layers = nn_bo_MCF_drop_layers.max['params']

#some error happened while running (it's not choosing iteration with max score), on inspection 6th iteration gave maximum score

#| 6         | 0.9821    | 0.3436    | 242.5     | 0.128     | 0.01001   | 38.11     | 2.088     | 1.357     | 0.1876    | 2.566e+03 | 0.683     | 3.283     |
#|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |

params_nn_MCF_drop_layers = {
    'activation': 0.3436, 
    'batch_size':242.5, 
    'dropout': 0.128, 
    'dropout_rate': 0.01001, 
    'epochs': 38.11, 
    'layers1':2.088, 
    'layers2':1.357, 
    'learning_rate':0.1876, 
    'neurons':2566, 
    'normalization': 0.683, 
    'optimizer':3.283
}
  
learning_rate = params_nn_MCF_drop_layers['learning_rate']
activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
               'elu', 'exponential', LeakyReLU,'relu']
params_nn_MCF_drop_layers['activation'] = activationL[round(params_nn_MCF_drop_layers['activation'])]

params_nn_MCF_drop_layers['batch_size'] = round(params_nn_MCF_drop_layers['batch_size'])
params_nn_MCF_drop_layers['epochs'] = round(params_nn_MCF_drop_layers['epochs'])
params_nn_MCF_drop_layers['layers1'] = round(params_nn_MCF_drop_layers['layers1'])
params_nn_MCF_drop_layers['layers2'] = round(params_nn_MCF_drop_layers['layers2'])
params_nn_MCF_drop_layers['neurons'] = round(params_nn_MCF_drop_layers['neurons'])

optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','Adam']
optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
             'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
             'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
             'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
params_nn_MCF_drop_layers['optimizer'] = optimizerD[optimizerL[round(params_nn_MCF_drop_layers['optimizer'])]]

params_nn_MCF_drop_layers

{'activation': 'relu',
 'batch_size': 242,
 'dropout': 0.128,
 'dropout_rate': 0.01001,
 'epochs': 38,
 'layers1': 2,
 'layers2': 1,
 'learning_rate': 0.1876,
 'neurons': 2566,
 'normalization': 0.683,
 'optimizer': <keras.optimizers.legacy.adadelta.Adadelta at 0x1dcbb8e9240>}

As we did for HCC, we fit the model to the training set and then make our predictions on the test set.

In [18]:
def nn_cl_fun_MCF_drop_layers():
    nn = Sequential()
    nn.add(Dense(params_nn_MCF_drop_layers['neurons'], input_dim=3000, activation=params_nn_MCF_drop_layers['activation']))
    if params_nn_MCF_drop_layers['normalization'] > 0.5:
        nn.add(BatchNormalization())
    for i in range(params_nn_MCF_drop_layers['layers1']):
        nn.add(Dense(params_nn_MCF_drop_layers['neurons'], activation=params_nn_MCF_drop_layers['activation']))
    if params_nn_MCF_drop_layers['dropout'] > 0.5:
        nn.add(Dropout(params_nn_MCF_drop_layers['dropout_rate'], seed=123))
    for i in range(params_nn_MCF_drop_layers['layers2']):
        nn.add(Dense(params_nn_MCF_drop_layers['neurons'], activation=params_nn_MCF_drop_layers['activation']))
    nn.add(Dense(1, activation='sigmoid'))
    nn.compile(loss='binary_crossentropy', optimizer=params_nn_MCF_drop_layers['optimizer'], metrics=['accuracy'])
    return nn

es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
nn = KerasClassifier(build_fn=nn_cl_fun_MCF_drop_layers, epochs=params_nn_MCF_drop_layers['epochs'], batch_size=params_nn_MCF_drop_layers['batch_size'],
                         verbose=0)
 
nn.fit(MCF_train_drop, MCF_y_train_drop, verbose=1)

Epoch 1/38
72/72 [==============================] - 26s 333ms/step - loss: 0.1482 - accuracy: 0.9442
Epoch 2/38
72/72 [==============================] - 27s 374ms/step - loss: 0.0382 - accuracy: 0.9882
Epoch 3/38
72/72 [==============================] - 32s 441ms/step - loss: 0.0168 - accuracy: 0.9956
Epoch 4/38
72/72 [==============================] - 25s 354ms/step - loss: 0.0073 - accuracy: 0.9984
Epoch 5/38
72/72 [==============================] - 25s 344ms/step - loss: 0.0040 - accuracy: 0.9992
Epoch 6/38
72/72 [==============================] - 24s 327ms/step - loss: 0.0034 - accuracy: 0.9992
Epoch 7/38
72/72 [==============================] - 24s 329ms/step - loss: 0.0020 - accuracy: 0.9997
Epoch 8/38
72/72 [==============================] - 23s 323ms/step - loss: 0.0018 - accuracy: 0.9997
Epoch 9/38
72/72 [==============================] - 23s 317ms/step - loss: 5.1349e-04 - accuracy: 1.0000
Epoch 10/38
72/72 [==============================] - 23s 320ms/step - loss: 0.0014 - ac

In [21]:
accuracy = accuracy_score(MCF_y_test_drop, nn.predict(MCF_test_drop))

print(accuracy)

136/136 [==============================] - 5s 39ms/step
0.9838150289017341


The Neural Network gets an accuracy of 98.4% on the test set. This is an improvement from our previous MLP Classifier.

### Summary of the models

Listed below are the optimal Neural Network models for both HCC1806 and MCF7:

HCC1806

{'activation': 'relu',  
'batch_size': 242,  
'dropout': 0.1279608146375314,  
'dropout_rate': 0.010014947837176436,  
'epochs': 38,  
'layers1': 2,  
'layers2': 1,  
'learning_rate': 0.18755133602737314,  
'neurons': 2566,  
'normalization': 0.6830131255680931,  
'optimizer': <keras.optimizers.legacy.adadelta.Adadelta at 0x1c292a579d0>}  

Test set accuracy: 95.61%

MCF7

{'activation': 'relu',  
'batch_size': 242,  
'dropout': 0.128,  
'dropout_rate': 0.01001,  
'epochs': 38,  
'layers1': 2,  
'layers2': 1,  
'learning_rate': 0.1876,  
'neurons': 2566,  
'normalization': 0.683,  
'optimizer': <keras.optimizers.legacy.adadelta.Adadelta at 0x17e0fde3040>}  
 
Test set Accuracy: 98.4%

Surpisingly, we get the same hyperparameters for both datasets.

---
---
### SmartSeq Neural Network Hyperparameter Tuning

We will also repeat the Neural Network Optimization procedure with SmartSeq. Runtime is much lower because the size of the training set is much smaller as opposed to DropSeq.

### HCC

In [24]:
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

# Load dataset
trainSet_HCC_smart = df_HCC

In [25]:
HCC_train_smart, HCC_test_smart = split_train_test(trainSet_HCC_smart, 0.2)
HCC_y_train_smart, HCC_y_test_smart = Label(HCC_train_smart,df_meta_HCC), Label(HCC_test_smart,df_meta_HCC)

In [19]:
# Create function
def nn_cl_bo2_HCC_smart(neurons, activation, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate):
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
                 'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
                 'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
                 'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
        
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']
        
    neurons = round(neurons)
    activation = activationL[round(activation)]
    optimizer = optimizerD[optimizerL[round(optimizer)]]
    batch_size = round(batch_size)
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)
        
    def nn_cl_fun():
        nn = Sequential()
        nn.add(Dense(neurons, input_dim=3000, activation=activation))
        if normalization > 0.5:
            nn.add(BatchNormalization())
        for i in range(layers1):
            nn.add(Dense(neurons, activation=activation))
        if dropout > 0.5:
            nn.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            nn.add(Dense(neurons, activation=activation))
        nn.add(Dense(1, activation='sigmoid'))
        nn.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
        return nn
        
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
    nn = KerasClassifier(build_fn=nn_cl_fun, epochs=epochs, batch_size=batch_size, verbose=0)
    
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, HCC_train_smart, HCC_y_train_smart, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    
    return score

In [22]:
params_nn2 ={
    'neurons': (3000, 100),
    'activation':(0, 9),
    'optimizer':(0,7),
    'learning_rate':(0.01, 1),
    'batch_size':(200, 1000),
    'epochs':(20, 100),
    'layers1':(1,3),
    'layers2':(1,3),
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}

#Run Bayesian Optimization
nn_bo_HCC_smart = BayesianOptimization(nn_cl_bo2_HCC_smart, params_nn2, random_state=111)
nn_bo_HCC_smart.maximize(init_points=25, n_iter=4)

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------
1/1 [==============================] - 0s 101ms/step
| 1         | 0.966     | 5.51      | 335.3     | 0.4361    | 0.2308    | 43.63     | 1.298     | 1.045     | 0.426     | 2.308e+03 | 0.3377    | 6.935     |
1/1 [==============================] - 0s 82ms/step
| 2         | 0.6492    | 2.14      | 265.0     | 0.6696    | 0.1864    | 41.94     | 1.932     | 1.237     | 0.08322   | 387.8     | 0.794     | 5.884     |
1/1 [==============================] - 0s 85ms/step
| 3         | 0.5069    | 7.337     | 992.8     | 0.5773    | 0.2441    | 53.71     | 1.055     | 1.908     | 0.1143    | 630.1     | 0.6977    | 3.957     |
1/1 [==============================] - 0s 139ms/step


ValueError: LBFGSB - one of the lower bounds is greater than an upper bound.

In [26]:
#| 19        | 0.9795    | 4.828     | 778.8     | 0.6616    | 0.2516    | 51.06     | 1.852     | 2.656     | 0.4743    | 621.9     | 0.01418   | 2.777     |
#|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
#params_nn_HCC_smart = nn_bo_HCC_smart.max['params']

params_nn_HCC_smart = {
 'activation': 4.828,
 'batch_size': 779,
 'dropout': 0.6615501030781817,
 'dropout_rate': 0.25163424049648575,
 'epochs': 51,
 'layers1': 2,
 'layers2': 3,
 'learning_rate': 0.4743187489952743,
 'neurons': 622,
 'normalization': 0.014176936309998611,
 'optimizer': 2.777}

learning_rate = params_nn_HCC_smart['learning_rate']
activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
               'elu', 'exponential', LeakyReLU,'relu']
params_nn_HCC_smart['activation'] = activationL[round(params_nn_HCC_smart['activation'])]

params_nn_HCC_smart['batch_size'] = round(params_nn_HCC_smart['batch_size'])
params_nn_HCC_smart['epochs'] = round(params_nn_HCC_smart['epochs'])
params_nn_HCC_smart['layers1'] = round(params_nn_HCC_smart['layers1'])
params_nn_HCC_smart['layers2'] = round(params_nn_HCC_smart['layers2'])
params_nn_HCC_smart['neurons'] = round(params_nn_HCC_smart['neurons'])

optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','Adam']
optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
             'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
             'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
             'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
params_nn_HCC_smart['optimizer'] = optimizerD[optimizerL[round(params_nn_HCC_smart['optimizer'])]]

params_nn_HCC_smart

{'activation': 'selu',
 'batch_size': 779,
 'dropout': 0.6615501030781817,
 'dropout_rate': 0.25163424049648575,
 'epochs': 51,
 'layers1': 2,
 'layers2': 3,
 'learning_rate': 0.4743187489952743,
 'neurons': 622,
 'normalization': 0.014176936309998611,
 'optimizer': <keras.optimizers.legacy.adadelta.Adadelta at 0x1dcc10a25f0>}

In [27]:
def nn_cl_fun_HCC_smart():
    nn = Sequential()
    nn.add(Dense(params_nn_HCC_smart['neurons'], input_dim=3000, activation=params_nn_HCC_smart['activation']))
    if params_nn_HCC_smart['normalization'] > 0.5:
        nn.add(BatchNormalization())
    for i in range(params_nn_HCC_smart['layers1']):
        nn.add(Dense(params_nn_HCC_smart['neurons'], activation=params_nn_HCC_smart['activation']))
    if params_nn_HCC_smart['dropout'] > 0.5:
        nn.add(Dropout(params_nn_HCC_smart['dropout_rate'], seed=123))
    for i in range(params_nn_HCC_smart['layers2']):
        nn.add(Dense(params_nn_HCC_smart['neurons'], activation=params_nn_HCC_smart['activation']))
    nn.add(Dense(1, activation='sigmoid'))
    nn.compile(loss='binary_crossentropy', optimizer=params_nn_HCC_smart['optimizer'], metrics=['accuracy'])
    return nn

es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
nn = KerasClassifier(build_fn=nn_cl_fun_HCC_smart, epochs=params_nn_HCC_smart['epochs'], batch_size=params_nn_HCC_smart['batch_size'],
                         verbose=0)
 
nn.fit(HCC_train_smart, HCC_y_train_smart, verbose=1)

Epoch 1/51
1/1 [==============================] - 2s 2s/step - loss: 240.2816 - accuracy: 0.4863
Epoch 2/51
1/1 [==============================] - 0s 39ms/step - loss: 1456.4105 - accuracy: 0.5137
Epoch 3/51
1/1 [==============================] - 0s 42ms/step - loss: 33.6049 - accuracy: 0.5479
Epoch 4/51
1/1 [==============================] - 0s 43ms/step - loss: 484.3991 - accuracy: 0.4863
Epoch 5/51
1/1 [==============================] - 0s 40ms/step - loss: 136.9702 - accuracy: 0.5137
Epoch 6/51
1/1 [==============================] - 0s 46ms/step - loss: 63.1885 - accuracy: 0.5137
Epoch 7/51
1/1 [==============================] - 0s 46ms/step - loss: 109.6094 - accuracy: 0.5137
Epoch 8/51
1/1 [==============================] - 0s 42ms/step - loss: 7.8097 - accuracy: 0.8288
Epoch 9/51
1/1 [==============================] - 0s 41ms/step - loss: 7.7621 - accuracy: 0.7260
Epoch 10/51
1/1 [==============================] - 0s 49ms/step - loss: 2.7201 - accuracy: 0.8767
Epoch 11/51
1/1 [=

In [29]:
accuracy = accuracy_score(HCC_y_test_smart, nn.predict(HCC_test_smart))

print(accuracy)

2/2 [==============================] - 0s 4ms/step
0.9722222222222222


The neural network gets an accuracy of 97.22% on HCC1806 SmartSeq. This is not a huge improvement compared to the normal MLP classifier (when K fold cross validation is implemented). Perhaps this is because the dataset is much smaller and the MLP classifier is sufficient to get good reuslts. 

---
### MCF

In [30]:
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

# Load dataset
trainSet_MCF_smart = df_MCF

In [31]:
MCF_train_smart, MCF_test_smart = split_train_test(trainSet_MCF_smart, 0.2)
MCF_y_train_smart, MCF_y_test_smart = Label(MCF_train_smart,df_meta_MCF), Label(MCF_test_smart,df_meta_MCF)

In [29]:
# Create function
def nn_cl_bo2_MCF_smart(neurons, activation, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate):
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
                 'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
                 'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
                 'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
        
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']
        
    neurons = round(neurons)
    activation = activationL[round(activation)]
    optimizer = optimizerD[optimizerL[round(optimizer)]]
    batch_size = round(batch_size)
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)
        
    def nn_cl_fun():
        nn = Sequential()
        nn.add(Dense(neurons, input_dim=3000, activation=activation))
        if normalization > 0.5:
            nn.add(BatchNormalization())
        for i in range(layers1):
            nn.add(Dense(neurons, activation=activation))
        if dropout > 0.5:
            nn.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            nn.add(Dense(neurons, activation=activation))
        nn.add(Dense(1, activation='sigmoid'))
        nn.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
        return nn
        
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
    nn = KerasClassifier(build_fn=nn_cl_fun, epochs=epochs, batch_size=batch_size, verbose=0)
    
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, MCF_train_smart, MCF_y_train_smart, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    
    return score

In [30]:
params_nn2 ={
    'neurons': (3000, 100),
    'activation':(0, 9),
    'optimizer':(0,7),
    'learning_rate':(0.01, 1),
    'batch_size':(200, 1000),
    'epochs':(20, 100),
    'layers1':(1,3),
    'layers2':(1,3),
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}

#Run Bayesian Optimization
nn_bo_MCF_smart = BayesianOptimization(nn_cl_bo2_MCF_smart, params_nn2, random_state=111)
nn_bo_MCF_smart.maximize(init_points=25, n_iter=4)

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------
2/2 [==============================] - 0s 18ms/step
| 1         | 1.0       | 5.51      | 335.3     | 0.4361    | 0.2308    | 43.63     | 1.298     | 1.045     | 0.426     | 2.308e+03 | 0.3377    | 6.935     |
2/2 [==============================] - 0s 5ms/step
| 2         | 1.0       | 2.14      | 265.0     | 0.6696    | 0.1864    | 41.94     | 1.932     | 1.237     | 0.08322   | 387.8     | 0.794     | 5.884     |
2/2 [==============================] - 0s 7ms/step
| 3         | 1.0       | 7.337     | 992.8     | 0.5773    | 0.2441    | 53.71     | 1.055     | 1.908     | 0.1143    | 630.1     | 0.6977    | 3.957     |
2/2 [==============================] - 0s 12ms/step
| 4 

ValueError: Input y contains NaN.

In [33]:
# params_nn_MCF_smart = nn_bo_MCF_smart.max['params']
#|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
#| 1         | 1.0       | 5.51      | 335.3     | 0.4361    | 0.2308    | 43.63     | 1.298     | 1.045     | 0.426     | 2.308e+03 | 0.3377    | 6.935     |

params_nn_MCF_smart = {
 'activation': 5.51,
 'batch_size': 371,
 'dropout': 0.29777470266523465,
 'dropout_rate': 0.2210307667883602,
 'epochs': 21,
 'layers1': 1,
 'layers2': 2,
 'learning_rate': 0.5032743170461116,
 'neurons': 2368,
 'normalization': 0.008929963256854245,
 'optimizer': 6.935}

learning_rate = params_nn_MCF_smart['learning_rate']
activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
               'elu', 'exponential', LeakyReLU,'relu']
params_nn_MCF_smart['activation'] = activationL[round(params_nn_MCF_smart['activation'])]

params_nn_MCF_smart['batch_size'] = round(params_nn_MCF_smart['batch_size'])
params_nn_MCF_smart['epochs'] = round(params_nn_MCF_smart['epochs'])
params_nn_MCF_smart['layers1'] = round(params_nn_MCF_smart['layers1'])
params_nn_MCF_smart['layers2'] = round(params_nn_MCF_smart['layers2'])
params_nn_MCF_smart['neurons'] = round(params_nn_MCF_smart['neurons'])

optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','Adam']
optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
             'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
             'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
             'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
params_nn_MCF_smart['optimizer'] = optimizerD[optimizerL[round(params_nn_MCF_smart['optimizer'])]]

params_nn_MCF_smart

{'activation': 'elu',
 'batch_size': 371,
 'dropout': 0.29777470266523465,
 'dropout_rate': 0.2210307667883602,
 'epochs': 21,
 'layers1': 1,
 'layers2': 2,
 'learning_rate': 0.5032743170461116,
 'neurons': 2368,
 'normalization': 0.008929963256854245,
 'optimizer': <keras.optimizers.legacy.ftrl.Ftrl at 0x1dca63fe980>}

In [34]:
def nn_cl_fun_MCF_smart():
    nn = Sequential()
    nn.add(Dense(params_nn_MCF_smart['neurons'], input_dim=3000, activation=params_nn_MCF_smart['activation']))
    if params_nn_MCF_smart['normalization'] > 0.5:
        nn.add(BatchNormalization())
    for i in range(params_nn_MCF_smart['layers1']):
        nn.add(Dense(params_nn_MCF_smart['neurons'], activation=params_nn_MCF_smart['activation']))
    if params_nn_MCF_smart['dropout'] > 0.5:
        nn.add(Dropout(params_nn_MCF_smart['dropout_rate'], seed=123))
    for i in range(params_nn_MCF_smart['layers2']):
        nn.add(Dense(params_nn_MCF_smart['neurons'], activation=params_nn_MCF_smart['activation']))
    nn.add(Dense(1, activation='sigmoid'))
    nn.compile(loss='binary_crossentropy', optimizer=params_nn_MCF_smart['optimizer'], metrics=['accuracy'])
    return nn

es = EarlyStopping(monitor='accuracy', mode='max', verbose=0, patience=20)
nn = KerasClassifier(build_fn=nn_cl_fun_MCF_smart, epochs=params_nn_MCF_smart['epochs'], batch_size=params_nn_MCF_smart['batch_size'],
                         verbose=0)
 
nn.fit(MCF_train_smart, MCF_y_train_smart, verbose=1)

Epoch 1/21
1/1 [==============================] - 1s 1s/step - loss: 251.9572 - accuracy: 0.0100
Epoch 2/21
1/1 [==============================] - 0s 394ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 3/21
1/1 [==============================] - 0s 356ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 4/21
1/1 [==============================] - 0s 317ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 5/21
1/1 [==============================] - 0s 295ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 6/21
1/1 [==============================] - 0s 273ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 7/21
1/1 [==============================] - 0s 283ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 8/21
1/1 [==============================] - 0s 280ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 9/21
1/1 [==============================] - 0s 261ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 10/21
1/1 [==============================] - 0s 276ms/step - loss: 0.0000e+00 - a

In [35]:
accuracy = accuracy_score(MCF_y_test_smart, nn.predict(MCF_test_smart))

print(accuracy)

2/2 [==============================] - 0s 15ms/step
1.0


Unsurprisingly, our neural network gets 100% accuracy on the test set. This was obtained by even our MLP Classifier.

### Summary of the Models

HCC1806

{'activation': 'selu',  
'batch_size': 779,  
'dropout': 0.6615501030781817,  
'dropout_rate': 0.25163424049648575,  
'epochs': 51,  
'layers1': 2,  
'layers2': 3,  
'learning_rate': 0.4743187489952743,  
'neurons': 622,  
'normalization': 0.014176936309998611,  
'optimizer': <keras.optimizers.legacy.adadelta.Adadelta at 0x26707a0a4a0>}

Test set Accuracy - 97.22%

MCF7

{'activation': 'elu',   
'batch_size': 371,  
'dropout': 0.29777470266523465,     
'dropout_rate': 0.2210307667883602,     
'epochs': 21,   
'layers1': 1,   
'layers2': 2,   
'learning_rate': 0.5032743170461116,    
'neurons': 2368,    
'normalization': 0.008929963256854245,  
'optimizer': <keras.optimizers.legacy.ftrl.Ftrl at 0x28d204eda50>}  

Test set Accuracy - 100%


This concludes the section on Multilayer Perceptron and Neural Networks.

---
---
---